File: Backend/src/utils/validators.js
const { body, validationResult } = require("express-validator");

const phoneValidator = body("phone_number")
  .optional({ checkFalsy: true })
  .trim()
  .matches(/^(\+251|0)?9\d{8}$/)
  .withMessage(
    "Phone number must be valid (Example: 0911223344 or +251911223344)",
  );

const validateRegister = [
  body("full_name")
    .trim()
    .notEmpty()
    .withMessage("Full name is required")
    .isLength({ min: 2, max: 100 })
    .withMessage("Full name must be between 2 and 100 characters"),
  body("email")
    .trim()
    .notEmpty()
    .withMessage("Email is required")
    .isEmail()
    .withMessage("Please provide a valid email")
    .normalizeEmail(),
  body("password")
    .notEmpty()
    .withMessage("Password is required")
    .isLength({ min: 6 })
    .withMessage("Password must be at least 6 characters"),
  body("phone_number")
    .optional({ checkFalsy: true })
    .trim()
    .isLength({ max: 20 })
    .withMessage("Phone number cannot exceed 20 characters")
    .custom((value) => {
      if (value && value.length > 0) {
        const phoneRegex = /^[0-9+\-\s()]+$/;
        if (!phoneRegex.test(value)) {
          throw new Error(
            "Phone number can only contain numbers, spaces, +, -, (, )",
          );
        }
      }
      return true;
    }),
];

const validateLogin = [
  body("email")
    .trim()
    .notEmpty()
    .withMessage("Email is required")
    .isEmail()
    .withMessage("Please provide a valid email"),
  body("password").notEmpty().withMessage("Password is required"),
];

const validateStartup = [
  body("title")
    .trim()
    .notEmpty()
    .withMessage("Startup title is required")
    .isLength({ min: 3, max: 200 })
    .withMessage("Title must be between 3 and 200 characters"),

  body("blurb")
    .trim()
    .notEmpty()
    .withMessage("Blurb is required")
    .isLength({ min: 10, max: 500 })
    .withMessage("Blurb must be between 10 and 500 characters"),

  body("pitch_link")
    .optional({ checkFalsy: true })
    .trim()
    .isURL()
    .withMessage("Please provide a valid URL"),
];

const validateProfileUpdate = [
  body("full_name")
    .optional({ checkFalsy: true })
    .trim()
    .isLength({ min: 2, max: 100 })
    .withMessage("Full name must be between 2 and 100 characters"),

  phoneValidator,

  body("email")
    .optional({ checkFalsy: true })
    .trim()
    .isEmail()
    .withMessage("Please provide a valid email")
    .normalizeEmail(),
];

const validateInterest = [
  body("startup_id")
    .notEmpty()
    .withMessage("Startup ID is required")
    .isInt({ min: 1 })
    .withMessage("Startup ID must be a valid integer"),

  body("message")
    .optional({ checkFalsy: true })
    .trim()
    .isLength({ max: 500 })
    .withMessage("Message cannot exceed 500 characters"),
];

const validateBookmark = [
  body("startup_id")
    .notEmpty()
    .withMessage("Startup ID is required")
    .isInt({ min: 1 })
    .withMessage("Startup ID must be a valid integer"),

  body("note")
    .optional({ checkFalsy: true })
    .trim()
    .isLength({ max: 500 })
    .withMessage("Note cannot exceed 500 characters"),
];

const validateIntro = [
  body("startup_id")
    .notEmpty()
    .withMessage("Startup ID is required")
    .isInt({ min: 1 })
    .withMessage("Startup ID must be a valid integer"),

  body("intro_text")
    .trim()
    .notEmpty()
    .withMessage("Intro text is required")
    .isLength({ min: 10, max: 1000 })
    .withMessage("Intro text must be between 10 and 1000 characters"),
];

const validatePasswordUpdate = [
  body("current_password")
    .notEmpty()
    .withMessage("Current password is required"),
  body("new_password")
    .notEmpty()
    .withMessage("New password is required")
    .isLength({ min: 6 })
    .withMessage("New password must be at least 6 characters"),
];

// Account deletion validation
const validateAccountDeletion = [
  body("password_confirmation")
    .notEmpty()
    .withMessage("Password confirmation is required"),
];

const validateRequest = (req, res, next) => {
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).json({
      success: false,
      message: "Validation error",
      errors: errors.array().map((err) => ({
        field: err.path,
        message: err.msg,
      })),
    });
  }

  next();
};

module.exports = {
  validateRegister,
  validateLogin,
  validateStartup,
  validateProfileUpdate,
  validateInterest,
  validateBookmark,
  validateIntro,
  validatePasswordUpdate,
  validateRequest,
};

File: lib/core/utils/validators.dart
class Validators {
  static String? validateEmail(String? email) {
    if (email == null || email.isEmpty) {
      return 'Email is required';
    }
    
    final emailRegex = RegExp(r'^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$');
    if (!emailRegex.hasMatch(email)) {
      return 'Enter a valid email address';
    }
    
    return null;
  }
  
  static String? validatePassword(String? password) {
    if (password == null || password.isEmpty) {
      return 'Password is required';
    }
    
    if (password.length < 6) {
      return 'Password must be at least 6 characters';
    }
    
    return null;
  }
  
  static String? validateFullName(String? name) {
    if (name == null || name.isEmpty) {
      return 'Full name is required';
    }
    
    if (name.length < 2) {
      return 'Name must be at least 2 characters';
    }
    
    if (name.length > 100) {
      return 'Name cannot exceed 100 characters';
    }
    
    return null;
  }
  
  static String? validatePhoneNumber(String? phone) {
    if (phone == null || phone.isEmpty) {
      return null; // Phone is optional
    }
    
    final phoneRegex = RegExp(r'^[0-9+\-\s()]{8,20}$');
    if (!phoneRegex.hasMatch(phone)) {
      return 'Enter a valid phone number';
    }
    
    return null;
  }
  
  static String? validateStartupTitle(String? title) {
    if (title == null || title.isEmpty) {
      return 'Startup title is required';
    }
    
    if (title.length < 3) {
      return 'Title must be at least 3 characters';
    }
    
    if (title.length > 200) {
      return 'Title cannot exceed 200 characters';
    }
    
    return null;
  }
  
  static String? validateStartupBlurb(String? blurb) {
    if (blurb == null || blurb.isEmpty) {
      return 'Startup description is required';
    }
    
    if (blurb.length < 10) {
      return 'Description must be at least 10 characters';
    }
    
    if (blurb.length > 500) {
      return 'Description cannot exceed 500 characters';
    }
    
    return null;
  }
  
  static String? validateUrl(String? url) {
    if (url == null || url.isEmpty) {
      return null; // URL is optional
    }
    
    final urlRegex = RegExp(r'^(http|https)://[^\s]+$');
    if (!urlRegex.hasMatch(url)) {
      return 'Enter a valid URL (http:// or https://)';
    }
    
    return null;
  }
  
  static String? validateIntroText(String? text) {
    if (text == null || text.isEmpty) {
      return 'Intro message is required';
    }
    
    if (text.length < 10) {
      return 'Message must be at least 10 characters';
    }
    
    if (text.length > 1000) {
      return 'Message cannot exceed 1000 characters';
    }
    
    return null;
  }
}